src/test/coinscachepair_tests.cpp

```diff
@@ -0,0 +1,186 @@
+    auto lit{nodes.begin()};
+    for (auto it{head.second.Next()}; it != nullptr; it = it->second.Next(/*clear_flags=*/false), ++lit) {
+        BOOST_CHECK_EQUAL(&(*lit), it);
+    }
+
+    // Check iterating through pairs is identical to iterating through a list
+    // Clear the flags during iteration
+    lit = nodes.begin();
+    for (auto it{head.second.Next()}; it != nullptr; it = it->second.Next(/*clear_flags=*/true), ++lit) {
+        BOOST_CHECK_EQUAL(&(*lit), it);
+    }
+    // Check that head is empty
+    BOOST_CHECK_EQUAL(head.second.Next(), nullptr);
+
+    // Delete the nodes from the list to make sure there are no dangling pointers
+    for (auto it{nodes.begin()}; it != nodes.end(); it = nodes.erase(it)) {
+        BOOST_CHECK_EQUAL(it->second.GetFlags(), 0);
+        BOOST_CHECK_EQUAL(it->second.Next(), nullptr);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(linked_list_iterate_erase)
+{
+    CoinsCachePair head;
+    auto nodes{CreatePairs(head)};
+
+    // Check iterating through pairs is identical to iterating through a list
+    // Erase the nodes as we iterate through, but don't clear flags
+    // The flags will be cleared by the CCoinsCacheEntry's destructor
+    auto lit{nodes.begin()};
+    for (auto it{head.second.Next()}; it != nullptr; it = it->second.Next(/*clear_flags=*/false), lit = nodes.erase(lit)) {
+        BOOST_CHECK_EQUAL(&(*lit), it);
+    }
+    // Check that head is empty
+    BOOST_CHECK_EQUAL(head.second.Next(), nullptr);
+}
+
+BOOST_AUTO_TEST_CASE(linked_list_random_deletion)
+{
+    CoinsCachePair head;
+    auto nodes{CreatePairs(head)};
+
+    // Create linked list head->n1->n2->n3->n4->nullptr
+    auto n1{nodes.begin()};
+    auto n2{std::next(n1)};
+    auto n3{std::next(n2)};
+    auto n4{std::next(n3)};
+
+    // Delete n2
+    // head->n1->n3->n4->nullptr
+    nodes.erase(n2);
+    // Check that n1 now points to n3, and n3 still points to n4
+    // Also check that flags were not altered
+    BOOST_CHECK_EQUAL(n1->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(n1->second.Next(), &(*n3));
+    BOOST_CHECK_EQUAL(n3->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(n3->second.Next(), &(*n4));
+
+    // Delete n1
+    // head->n3->n4->nullptr
+    nodes.erase(n1);
+    // Check that head now points to n3, and n3 still points to n4
+    // Also check that flags were not altered
+    BOOST_CHECK_EQUAL(n3->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(head.second.Next(), &(*n3));
+    BOOST_CHECK_EQUAL(n3->second.Next(), &(*n4));
+
+    // Delete n4
+    // head->n3->nullptr
+    nodes.erase(n4);
+    // Check that head still points to n3, and n3 points to nullptr
+    // Also check that flags were not altered
+    BOOST_CHECK_EQUAL(n3->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+    BOOST_CHECK_EQUAL(head.second.Next(), &(*n3));
+    BOOST_CHECK_EQUAL(n3->second.Next(), nullptr);
+
+    // Delete n3
+    // head->nullptr
+    nodes.erase(n3);
```

Comment:
It just checks that removing the last element sets head to `nullptr`. I suppose just for completeness, since this case is the same as when we delete `n4` above.

Reply:
Updated comment.